http-server/src/main/java/io/micronaut/http/server/RouteExecutor.java

```diff
@@ -667,11 +676,126 @@ private Mono<MutableHttpResponse<?>> processPublisherBody(PropagatedContext prop
             propagatedContext
         ).contextWrite(cv -> ReactorPropagation.addPropagatedContext(cv, propagatedContext).put(ServerRequestContext.KEY, request));
 
-        return Mono.<MutableHttpResponse<?>>just(response
-            .header(HttpHeaders.TRANSFER_ENCODING, "chunked")
-            .header(HttpHeaders.CONTENT_TYPE, mediaType)
-            .body(ReactivePropagation.propagate(propagatedContext, bodyPublisher)))
-            .contextWrite(context -> ReactorPropagation.addPropagatedContext(context, propagatedContext).put(ServerRequestContext.KEY, request));
+        return new StreamingDataProcessor(bodyPublisher).chunkedResponsePublisher(propagatedContext, request, response, routeInfo, mediaType);
+    }
+
+    /**
+     * Stateful processor for writing a streaming data HTTP response.
+     * <p>
+     * The first signal from the underlying body publisher is inspected before writing the response
+     * in order to check for potential errors and a matching error route. If the publisher signals
+     * an immediate error before writing any data and there is a matching error route, then the
+     * error will be propagated outward so that the error route will be invoked. Otherwise, an
+     * appropriate {@link MutableHttpResponse} will be published that is able to stream the body
+     * publisher as a chunked HTTP response.
+     */
+    private class StreamingDataProcessor implements Consumer<FluxSink<Object>> {
+
+        private final Flux<Object> bodyPublisher;
+
+        private boolean firstRequest = true;
+
+        private boolean bufferWritten = false;
+
+        private FluxSink<Object> bodySink;
+
+        private boolean subscribed = false;
+
+        private Object bufferedFirstValue = null;
+
+        private Throwable bufferedImmediateError = null;
+
+        private final StreamSubscriber subscriber = new StreamSubscriber();
+
+        private boolean isComplete;
+
+        private boolean isEmpty;
+
+        private StreamingDataProcessor(Flux<Object> bodyPublisher) {
+            this.bodyPublisher = bodyPublisher;
+        }
+
+        @SuppressWarnings("unchecked")
+        private CorePublisher<MutableHttpResponse<?>> chunkedResponsePublisher(PropagatedContext propagatedContext,
```

Comment:
this needs to be thread safe

Reply:
@yawkat Maybe I'm missing something...how is this method not thread-safe?

Reply:
for example the second accept can be called concurrently with any of the methods of StreamSubscriber

Reply:
Oh, you mean the publisher itself. That is a good point. I don't believe that will currently happen just because of the way the publisher is being consumed by the rest of the pipeline, but that is definitely too unsafe an assumption.